File: api/skilltree2-progress.js
import { MongoClient } from 'mongodb';
import { getAuthFromRequest } from './auth-shared.js';

const uri = process.env.SKILLTREE2_MONGODB_URI;
const dbName = process.env.SKILLTREE2_DB_NAME || 'skilltree2';

let cachedClient = null;

async function getDb() {
    if (cachedClient && cachedClient.topology && cachedClient.topology.isConnected()) {
        return cachedClient.db(dbName);
    }
    const client = new MongoClient(uri);
    await client.connect();
    cachedClient = client;
    return client.db(dbName);
}

function sendJson(res, status, body) {
    res.statusCode = status;
    res.setHeader('Content-Type', 'application/json; charset=utf-8');
    res.setHeader('Cache-Control', 'no-store');
    res.end(JSON.stringify(body));
}

async function readJsonBody(req) {
    if (req.body && typeof req.body === 'object') return req.body;
    if (typeof req.body === 'string') {
        try { return JSON.parse(req.body); } catch { return null; }
    }
    const raw = await new Promise((resolve, reject) => {
        let data = '';
        req.on('data', (chunk) => (data += chunk));
        req.on('end', () => resolve(data));
        req.on('error', reject);
    });
    if (!raw) return null;
    try { return JSON.parse(raw); } catch { return null; }
}

export default async function handler(req, res) {
    if (!uri) {
        return sendJson(res, 500, { error: 'SKILLTREE2_MONGODB_URI not configured' });
    }

    const auth = getAuthFromRequest(req);
    if (!auth || !auth.userId) {
        return sendJson(res, 401, { error: 'Authentication required' });
    }
    const userId = auth.userId;

    const method = req.method || 'GET';

    try {
        const db = await getDb();
        const col = db.collection('progress');

        if (method === 'GET') {
            const doc = await col.findOne({ _id: userId });
            if (doc) {
                const { _id, ...data } = doc;
                return sendJson(res, 200, data);
            }
            return sendJson(res, 200, { topicStatus: {}, notes: {}, savedAt: null });
        }

        if (method === 'POST') {
            const payload = await readJsonBody(req);
            if (!payload || typeof payload !== 'object') {
                return sendJson(res, 400, { error: 'Invalid JSON' });
            }

            const { topicStatus, notes } = payload;
            const doc = {
                topicStatus: topicStatus || {},
                notes: notes || {},
                savedAt: new Date().toISOString(),
            };

            await col.updateOne(
                { _id: userId },
                { $set: doc },
                { upsert: true },
            );

            return sendJson(res, 200, { ok: true, savedAt: doc.savedAt });
        }

        return sendJson(res, 405, { error: 'Method Not Allowed' });
    } catch (err) {
        console.error('skilltree2-progress error:', err);
        return sendJson(res, 500, { error: 'Database error' });
    }
}

File: api/skilltree2_progress.py
"""
Skill Tree 2.0 progress API — MongoDB CRUD for per-user topic progress.
Port of api/skilltree2-progress.js
"""

import os
from datetime import datetime, timezone
from flask import Flask, request
from pymongo import MongoClient
from api.auth_utils import get_auth_from_request, send_json

app = Flask(__name__)

MONGODB_URI = os.environ.get('SKILLTREE2_MONGODB_URI', '')
DB_NAME = os.environ.get('SKILLTREE2_DB_NAME', 'skilltree2')

_client = None


def get_db():
    global _client
    if not MONGODB_URI:
        return None
    if _client is not None:
        try:
            _client.admin.command('ping')
            return _client[DB_NAME]
        except Exception:
            _client = None
    _client = MongoClient(MONGODB_URI, serverSelectionTimeoutMS=10000)
    _client.admin.command('ping')
    return _client[DB_NAME]


@app.route('/api/skilltree2-progress', methods=['GET', 'POST'])
def skilltree2_progress():
    if not MONGODB_URI:
        return send_json({'error': 'SKILLTREE2_MONGODB_URI not configured'}, 500)

    auth = get_auth_from_request(request)
    if not auth or not auth.get('userId'):
        return send_json({'error': 'Authentication required'}, 401)

    user_id = auth['userId']

    try:
        db = get_db()
        col = db.progress

        if request.method == 'GET':
            doc = col.find_one({'_id': user_id})
            if doc:
                doc.pop('_id', None)
                return send_json(doc)
            return send_json({'topicStatus': {}, 'notes': {}, 'savedAt': None})

        if request.method == 'POST':
            payload = request.get_json(silent=True)
            if not payload or not isinstance(payload, dict):
                return send_json({'error': 'Invalid JSON'}, 400)

            doc = {
                'topicStatus': payload.get('topicStatus', {}),
                'notes': payload.get('notes', {}),
                'savedAt': datetime.now(timezone.utc).isoformat(),
            }
            col.update_one({'_id': user_id}, {'$set': doc}, upsert=True)
            return send_json({'ok': True, 'savedAt': doc['savedAt']})

    except Exception as e:
        print(f'skilltree2-progress error: {e}')
        return send_json({'error': 'Database error'}, 500)

    return send_json({'error': 'Method Not Allowed'}, 405)
